pkg/reconciler/taskrun/taskrun.go

```diff
@@ -333,10 +333,27 @@ func (c *Reconciler) prepare(ctx context.Context, tr *v1beta1.TaskRun) (*v1beta1
 		return nil, nil, controller.NewPermanentError(err)
 	}
 
-	if err := validateTaskSpecRequestResources(ctx, taskSpec); err != nil {
-		logger.Errorf("TaskRun %s taskSpec request resources are invalid: %v", tr.Name, err)
+	if isResourcesRequirementsInStepLevel(taskSpec) && isResourcesRequirementsInTaskLevel(taskSpec) {
+		err := fmt.Errorf("TaskRun %s can't be configured with both step-level and task-level resources requirements", tr.Name)
 		tr.Status.MarkResourceFailed(podconvert.ReasonFailedValidation, err)
 		return nil, nil, controller.NewPermanentError(err)
+	} else if isResourcesRequirementsInStepLevel(taskSpec) {
```

Comment:
We should apply validation in the webhook as well as the reconciler (the reason is to return errors as early as possible -- for example an invalid task should be rejected on creation, not when it's run in a taskrun). I'd suggest:

- in [task_validation.go](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/task_validation.go), ensure that the task does not have both step-level and task-level resource requirements
- in [taskrun_validation.go](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/taskrun_validation.go), make the same check as for tasks
- Here in the reconciler code (probably as a helper function in [validate_resources.go](https://github.com/tektoncd/pipeline/blob/c938ee5667fb64ddcf6e6c26e7732b68dde82104/pkg/reconciler/taskrun/validate_resources.go)), validate that the taskrun resource requirements are compatible with the task resource requirements.

Reply:
thanks Lee. let me only keep "validation" logic in the reconciler (removing the resources update part) and add validation in webhook.

Reply:
closed by moving changes on reconciler into the following PR